src/boot/fwupd.c

```diff
@@ -0,0 +1,630 @@
+                        /* SearchKey= */ NULL,
+                        &n_handles,
+                        &handles);
+        if (err != EFI_SUCCESS)
+                return log_error_status(err, "Failed to locate graphics output protocol handles: %m");
+        if (n_handles == 0 || !handles)
+                return log_error_status(EFI_UNSUPPORTED, "No graphics output protocol handles found for UX capsule");
+
+        FOREACH_ARRAY(h, handles, n_handles) {
+                EFI_GRAPHICS_OUTPUT_PROTOCOL *proto;
+
+                err = BS->OpenProtocol(*h,
+                                       MAKE_GUID_PTR(EFI_GRAPHICS_OUTPUT_PROTOCOL),
+                                       (void **)&proto,
+                                       image,
+                                       /* ControllerHandler= */ NULL,
+                                       EFI_OPEN_PROTOCOL_GET_PROTOCOL);
+                if (err != EFI_SUCCESS)
+                        return log_error_status(err, "Failed to get graphics output protocol: %m");
+
+                *ret_mode = mode = proto->mode->mode;
+                found = 1;
+                return EFI_SUCCESS;
+        }
+
+        found = 0;
+        return EFI_UNSUPPORTED;
+}
+
+static EFI_DEVICE_PATH *device_path_find_type(EFI_DEVICE_PATH **dp, uint8_t type) {
+        EFI_DEVICE_PATH *next;
+
+        assert(dp);
+        assert(*dp);
+
+        for (;;) {
+                next = device_path_next_node(*dp);
+
+                if ((*dp)->Type == type)
+                        break;
+
+                *dp = next;
+        }
+
+        return next;
+}
+
+static EFI_DEVICE_PATH *device_path_instance(EFI_DEVICE_PATH **device_path, size_t *ret_size) {
+        EFI_DEVICE_PATH *head, *dp;
+
+        assert(device_path);
+        assert(ret_size);
+
+        if (!*device_path) {
+                *ret_size = 0;
+                return NULL;
+        }
+
+        dp = head = *device_path;
+
+        /* Skip through all subtypes until the last node and return it */
+        *device_path = device_path_find_type(&dp, END_DEVICE_PATH_TYPE);
+        if (dp->SubType == END_ENTIRE_DEVICE_PATH_SUBTYPE)
+                *device_path = NULL;
+
+        *ret_size = ((uint8_t *) dp) - ((uint8_t *) head);
+
+        return head;
+}
+
+static EFI_STATUS device_path_find_index_of_type(EFI_DEVICE_PATH *dp, size_t *ret_index) {
+        size_t index = 0;
+
+        assert(ret_index);
+
+        if (!dp) {
+                *ret_index = 0;
+                return EFI_SUCCESS;
+        }
+
+        for (;;) {
+                if (device_path_is_end(dp))
+                        return log_error_status(EFI_INVALID_PARAMETER, "Device path for is empty");
+                if (dp->Type == MEDIA_DEVICE_PATH && dp->SubType == MEDIA_FILEPATH_DP)
+                        break;
+
+                index++;
+                dp = device_path_next_node(dp);
+        }
+
+        *dp = DEVICE_PATH_END_NODE;
```

Comment:
you are modifying the device path here, but the name of this function suggests this was about finding something, i.e. an operation that doesn't modify